### ISSUE TITLE
Hermes direct debugging not working as InstanceImpl does not call DevSupportManager.StartInspector.

### ISSUE THREAD

**Issue by nichamp** [microsoft/react-native-windows#9407]
### Problem Description

When using Hermes with a default project and the configuration described in https://github.com/microsoft/react-native-windows/blob/main/docs/inspector.md the RNW instance does not appear in the targets list for edge://inspect (or chrome://inspect) so it isn't possible to debug code with the built-in Hermes engine.

Under a debugger, one can observe that `m_devSettings->jsiEngineOverride == JSIEngineOverride::Default` in `InstanceImpl::InstanceImpl` (and the destructor too). If one steps into the block with the debugger, the instance will appear in the devices list (perhaps on refreshing by clicking `Configure` and `Done` again). Even if one sets InstanceSettings.JSIEngineOverride to JSIEngine.Hermes, the issue still occurs as `DevSettings.jsiEngineOverride` does not appear to get set anywhere to reflect the InstanceSettings' value. So it appears that this obsolete value should be removed (wuthout breaking non-Hermes scenarios) or set to match so that debugging works.

### Steps To Reproduce

1. Create a new RNW project using `--useHermes`
1. Modify App.cpp to set UseWebDebugger=false and UseDirectDebugger=true (or use Ctrl+Shift+D later)
1. Run `npx react-native windows` (or build+deploy with VS and start the metro server manually)
1. Open `edge://inspect` or `chrome://inspect` and on the Devices tab, click `Configure` to add `localhost:8081`
1. Observe that the target doesn't appear in the devices list.

### Expected Results

Debug builds (which include HermesInspector.dll) of the app using the Hermes runtime and are connected to a Metro bundler server and have developer support and direct debugging enabled should appear in debugging tools automatically with no additional configuration beyond that descrbed in https://github.com/microsoft/react-native-windows/blob/main/docs/inspector.md

### CLI version

6.3.1

### Environment

```markdown
System:
    OS: Windows 10 10.0.19042
    CPU: (24) x64 AMD Ryzen Threadripper PRO 3945WX 12-Cores
    Memory: 37.71 GB / 63.86 GB
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.15 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.18362.0, 10.0.19041.0, 10.0.22000.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.11.32106.194 (Visual Studio Enterprise 2019)
  Languages:
    Java: Not Found
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2
    react-native: 0.66.4 => 0.66.4
    react-native-windows: 0.66.7 => 0.66.7
  npmGlobalPackages:
    *react-native*: Not Found
```

### Target Platform Version

10.0.19041

### Visual Studio Version

Visual Studio 2019

### Build Configuration

Debug

**Comment by nichamp** [comment]
An additional note is that even with the fix, source mapping (for TS or JS) doesn't seem to work correctly unless I add ` m_devSettings->inlineSourceMap = true;` in the constructor too. I think this either needs to be the default with Hermes or maybe to have it exposed with InstanceSettings (maybe to allow arbitrary URL modifications too if desired). If there is an additional configuration step that I am missing otherwise, it should probably be added to https://github.com/microsoft/react-native-windows/blob/main/docs/inspector.md

**Comment by NickGerleman** [comment]
cc @mganandraj (also on Teams, if you two need to chat offline)

**Comment by mganandraj** [comment]
I've created a PR to ensure we start the inspector: https://github.com/microsoft/react-native-windows/pull/9426

**Comment by mganandraj** [comment]
@nichamp Thanks for sharing the issues .. 
I'm also seeing the issues with the source map. Need to investigate what changed recently.
And, there are known issues with breaking on first line. 
Adding @tudorms

**Comment by nichamp** [comment]
@mganandraj I suggest adding `inlineSourceMap = true` with your fix since that does work for me or is there a reason we should avoid that?

Are there already issues for `Enable Break on First Line` and hangs on reload?

**Comment by acoates-ms** [comment]
I would expect `Enable Break on First Line` to "hang" on instance reload (which will happen when you initially check it in the devmenu), until you connect the debugger and continue.   This is so that you can debug the startup JS that would get run before you have time to attach? Is that the hang you are seeing?

The sourcemap issue is likely an issue around your metro config -- It should be querying metro for your sourcemap. -- Using inline source map should work ok on dev builds, it puts the sourcemap information into your bundle file.  But having the sourcemap be stored separately is I think the default in core react-native, and I see no reason why react-native-windows would want to be different than core for that.

**Comment by nichamp** [comment]
> I would expect Enable Break on First Line to "hang" on instance reload (which will happen when you initially check it in the devmenu), until you connect the debugger and continue. This is so that you can debug the startup JS that would get run before you have time to attach? Is that the hang you are seeing?

`Enable Break on First Line` itself doesn't work so doesn't appear to be the problem. I am seeing the hang less frequently today than when I left the comment. It does seem to occur if one reloads while the debugger is attached but I thought I had observed it even after disconnecting. I guess these issues need some more concrete observations on my part to identify their repro conditions.

I have also observed the metro server sometimes crashing after trying to connect again after successfully reloading the bundle though I haven't yet looked at whether Metro has any issue tracking it:
 
![image](https://user-images.githubusercontent.com/46539320/150887345-0da1fa5c-145f-491d-a5e9-b4ee1879864b.png)

> The sourcemap issue is likely an issue around your metro config -- It should be querying metro for your sourcemap. -- Using inline source map should work ok on dev builds, it puts the sourcemap information into your bundle file. But having the sourcemap be stored separately is I think the default in core react-native, and I see no reason why react-native-windows would want to be different than core for that.

I was seeing this even with a vanilla project create with the init script. Which config parameters do you think should be added? As I mentioned earlier, it seems like devtools attempts to get the map (index.map) for the wrong file (index.js) instead of index.bundle, so having RNW retrieve the bundle with the query param for the inline map was the obvious solution that I found at the time to make it work. It would only be for dev builds using Hermes direct debugging with a Metro bundler since the server's inspector proxy is needed for debugging (and so debugging a bundle file is likely not an option I am assuming)

**Comment by mganandraj** [comment]
The issues with loading source maps when debugging playground app seems to be because of the source mapping URL embedded in the bundle ... This is how the embedded sourceMappingUrl verb looks like:

},683,[53,1,507],"..\\@react-native-windows\\tester\\src\\js\\components\\RNTesterEmptyBookmarksState.windows.js");
__r(0);
//# sourceMappingURL=//localhost:8081/Samples/rntester.map?platform=windows&dev=true&hot=true&inlineSourceMap=false
//# sourceURL=http://localhost:8081/Samples/rntester.bundle?platform=windows&dev=true&hot=true&inlineSourceMap=false

Note that "http:" is missing from the sourceMappingUrl in the bundle served by metro .. I need to dig into the metro code to see what is going wrong here ..

devtools gets the sourceMappingUrl from the "ScriptParsed" inspector event raised by hermes .. hermes gets the sourceMap path from the JavaScript bundle passed to it .. It is unlikely that devtools tries to get the map from wrong file .. 
@nichamp Could you try to open the bundle in a text editor and check the sourceMappingUrl embedded in it ? When served from metro, you could view the bundle by typing the metro http endpoint, for e.g. "http://localhost:8081/Samples/rntester.bundle?platform=windows&dev=true&hot=true&inlineSourceMap=false" for playground

**Comment by nichamp** [comment]
@mganandraj with the vanilla RNW app I see the following as you found:

```
//# sourceMappingURL=//localhost:8081/index.map?platform=windows&dev=true&hot=false&inlineSourceMap=false
//# sourceURL=http://localhost:8081/index.bundle?platform=windows&dev=true&hot=false&inlineSourceMap=false
```

**Comment by acoates-ms** [comment]
@nichamp can you verify if making the change from https://github.com/facebook/metro/pull/763 locally fixes your issue.

**Comment by nichamp** [comment]
@acoates-ms with that fix it does change the bundle to

```
//# sourceMappingURL=http://localhost:8081/index.map?platform=windows&dev=true&hot=true&inlineSourceMap=false
//# sourceURL=http://localhost:8081/index.bundle?platform=windows&dev=true&hot=true&inlineSourceMap=false
```

but source maps still fail:

![image](https://user-images.githubusercontent.com/46539320/151075136-cb990632-deab-4b5f-9680-a3acc05feead.png)

**Comment by acoates-ms** [comment]
Interesting.  I think that might be an edge/chrome issue.  Although it's also happening in vscode.
The URL in the sourcemappingURL is correct, and if you navigate manually to it, metro returns the sourcemap correctly.

I've been using flipper to debug, and flipper does seem to work for me.  Not sure what its doing special since I'm pretty sure its just an electron app, which I would have thought would work the same as edge/chrome.
https://fbflipper.com/docs/getting-started/index/

**Comment by RedMickey** [comment]
> Although it's also happening in vscode.

Hi @acoates-ms, could you please describe a bit the incorrect behavior in VS Code if you use React Native Tools extension?

When implementing support of React Native Windows, we also faced this peculiarity (`//# sourceMappingURL=//localhost:8081/Samples/rntester.map...`) and added a [special handler](https://github.com/microsoft/vscode-react-native/blob/master/src/debugger/sourceMap.ts#L150-L155) for this case.

**Comment by acoates-ms** [comment]
@RedMickey 
So I was testing this out in the react-native-windows repo.  My first attempt at using the vscode extension actually failed due to the extension not working with canary builds of react-native, which react-native-windows's main branch uses.
https://github.com/microsoft/vscode-react-native/pull/1734 should fix that.

It does look like your special case logic in the extension does allow debugging using that extension.  -- I'm hoping to get a fix in metro so that that special case is no longer needed (although I suppose it'll stay around for older versions of RN)

**Comment by nichamp** [comment]
> Interesting. I think that might be an edge/chrome issue. Although it's also happening in vscode. The URL in the sourcemappingURL is correct, and if you navigate manually to it, metro returns the sourcemap correctly.
> 
> I've been using flipper to debug, and flipper does seem to work for me. Not sure what its doing special since I'm pretty sure its just an electron app, which I would have thought would work the same as edge/chrome. https://fbflipper.com/docs/getting-started/index/

It's good to know that Flipper (didn't for me, but I probably wasn't using it correctly) and the VS Code plugin will work, but do you anticipate finding a fix for the Edge/Chrome devtools too? In the interim, it may be worth updating the RNW documentation to suggest Flipper/vscode-react-native.

**Comment by tudorms** [comment]
@nichamp it looks like the debugger not starting issue is fixed in main and in 0.67.1; can you confirm it works on your side as well?

We may need a follow-up for the source maps not loading in Edge / Chrome but I suspect that behavior is coming from upstream (repros in Android/iOS).

**Comment by nichamp** [comment]
My local compile-time patching workarounds for this issue and the source mapping are what we are using presently. I suspect it isn't worth updating to 0.67.1+ (unless there are other helpful fixes too) until the source mapping issue is addressed too so I can remove the patching for the entire file. Otherwise, I would have to update the patch for the updated RNW to still fix source maps.

**Comment by tudorms** [comment]
Is that locally applying acoates' Metro PR or a different local patch for source maps?

**Comment by nichamp** [comment]
His fix didn't address edge://inspect as discussed above. I am using the fix I described from when I filed the issue by forcing inline source maps in the scenario of using Hermes direct debugging

**Comment by acoates-ms** [comment]
@tudorms , it looks like you forced inlineSourceMaps to be false for hermes way back in this commit https://github.com/microsoft/react-native-windows/commit/60e137c319cb6e50a9efd5f7ffc80353cb980123, with the comment "* Disable inline sourcemaps for Hermes (it doesn't handle them well)".   Any idea if that still applies.  I agree that generally for the dev scenario, if we can load a bundle with inline sourcemaps it avoids a class of issues with trying to find the sourcemaps.  So if there isn't an issue with that, we should probably remove the line that turns off inline sourcemaps for hermes.

**Comment by tudorms** [comment]
Next build of 0.67 (0.67.3 due on Feb 28) will include the fix for InlineSourceMap.

**Comment by acoates-ms** [comment]
I think the InlineSourceMap fixes should unblock the scenario here.

**Comment by mganandraj** [comment]
Debugging into chrome, it turns out that the regression in loading sourcemap from url is introduced by this change : https://chromium-review.googlesource.com/c/chromium/src/+/3227594/
The change enforces that the resources be fetched from the storage sandbox identified by the web content in the chrome tab. In the standalone inspector instance that we launch through chrome:\\inspect, the web contents is null (A null check is added in a subsequent CR).

node debugging still works because file:// scheme is still supported.

I don't understand how it works with flipper. Flipper creates an electron 'webview' to load the inspector. Accordint to documentation, electron webview creates a standalone sandbox chrome instance, and it is very much a chromium webview. I expect the same issue to happen within webview, but i can't debug there yet. It is possible that electron version used by flipper is old enough to not have the above change.
And, if not so, chromium webviews are undergoing similar refactorings and the issue might soon appear there too.

Another note is that this issue is not windows specific. It affects Android/iOS too.

It looks like, there are fundamental security issues with loading source maps from localhost urls. Going forward it looks like inline source maps may be only option.

@tudorms i remember there used to be memory management issues with Hermes when loading inline source maps. The memory allocations used to grow to GBs. I hope that is fixed now.

**Comment by tudorms** [comment]
Thanks for the detailed investigation. Yes, the issue with inline source maps was fixed in upstream a while back, but we never got around to removing the hack: https://github.com/facebook/hermes/commit/093d15fb

**Comment by aeulitz** [comment]
It appears that as of commit 95638a in RNW main, the titular issue is fixed. Using the RNW playground apps (both Win32 and UWP and Chrome/Edge inspect, I verified that

1. the debug target is listed and that
2. source code can be loaded via Ctrl+P

Commit 95638a fixes an issue where re-loading packages does not remove debug targets from the Chrome/Edge inspect list. I did notice hangs in the playground apps and in vanilla project apps (generated via the CLI) when re-loading after connecting with a debugger to the previous instance. I suggest we create a dedicated issue for this.
